# 
The problem is to predict the safety of the car. In this project, I build a Decision Tree Classifier to predict the safety of the car. I implement Decision Tree Classification with Python and Scikit-Learn. I have used the Car Evaluation Data Set for this project, downloaded from the UCI Machine Learning Repository website.

#  Python libraries

• Numpy – It provides a fast numerical array structure and operating functions.

• pandas – It provides tools for data storage, manipulation and analysis tasks.

• Scikit-Learn – The required machine learning library in Python.

• Matplotlib – It is the basic plotting library in Python. It provides tools for making plots.

In [6]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [8]:
data = pd.read_csv(r"C:\Users\ASHOK\Downloads\car.data")


# Exploratoty Data Analysis

Now, I will explor the data to gain about the data.

In [10]:
print(data.shape)

(1727, 7)


# Veiw top 5 raws of dataset

In [11]:
print(data.head())

   vhigh vhigh.1  2 2.1  small   low  unacc
0  vhigh   vhigh  2   2  small   med  unacc
1  vhigh   vhigh  2   2  small  high  unacc
2  vhigh   vhigh  2   2    med   low  unacc
3  vhigh   vhigh  2   2    med   med  unacc
4  vhigh   vhigh  2   2    med  high  unacc


# Recolumn Names

In [17]:
col_names = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']


data.columns = col_names

col_names

['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']

In [22]:
# Let's again preveiw dataset
data.head()

buying  maint doors persons lug_boot safety  class
0  vhigh  vhigh     2       2    small    med  unacc
1  vhigh  vhigh     2       2    small   high  unacc
2  vhigh  vhigh     2       2      med    low  unacc
3  vhigh  vhigh     2       2      med    med  unacc
4  vhigh  vhigh     2       2      med   high  unacc

# veiw summary of dataset

In [18]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1727 entries, 0 to 1726
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   buying    1727 non-null   object
 1   maint     1727 non-null   object
 2   doors     1727 non-null   object
 3   persons   1727 non-null   object
 4   lug_boot  1727 non-null   object
 5   safety    1727 non-null   object
 6   class     1727 non-null   object
dtypes: object(7)
memory usage: 94.6+ KB
None


# Veiw descriptive statistics

In [19]:
print(data.describe())

       buying maint doors persons lug_boot safety  class
count    1727  1727  1727    1727     1727   1727   1727
unique      4     4     4       3        3      3      4
top      high  high     3       4      med   high  unacc
freq      432   432   432     576      576    576   1209


# Summary of variables:

There are 7 variables in the dataset.All the variables of categorical datatype.

There are given  buying, maint, doors, persons, lug_boot, safety and  class.

Class is the target variable.

# Explor class variable

In [23]:
#How many class lables are present
data["class"].value_counts()

unacc    1209
acc       384
good       69
vgood      65
Name: class, dtype: int64

# Missing values in variables

In [24]:
#Checking missing values in variables.
data.isnull().sum()

buying      0
maint       0
doors       0
persons     0
lug_boot    0
safety      0
class       0
dtype: int64

# Declear feature variables and target variable.

In [25]:
X = data.drop(["class"], axis = 1)
y = data["class"]

# Split data into separate training and test data

In [28]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.33, random_state = 45)

In [30]:
# check  the shape of X_train and X_test
X_train.shape, X_test.shape

((1157, 6), (570, 6))

# Feature Engineering.

Feature engineering is the process of transforming raw data into useful features that helps to understand our model better and increasing it's predictive power.

First, I will check the data type variables again

In [33]:
X_train.dtypes

buying      object
maint       object
doors       object
persons     object
lug_boot    object
safety      object
dtype: object

In [34]:
X_train.head()

buying  maint  doors persons lug_boot safety
1478    low   high      4    more      med    low
47    vhigh  vhigh      3    more      med    low
1230    med    low      3       4      big    med
909     med  vhigh      3    more    small    med
308   vhigh    med  5more       4      med    low

# Encode categorical variable

In [37]:
import category_encoders as ce
# encode variables with ordinal encoding

encoder = ce.OrdinalEncoder(cols=['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety'])


X_train = encoder.fit_transform(X_train)

X_test = encoder.transform(X_test)

In [38]:
X_train.head()

buying  maint  doors  persons  lug_boot  safety
1478       1      1      1        1         1       1
47         2      2      2        1         1       1
1230       3      3      2        2         2       2
909        3      2      2        1         3       2
308        2      4      3        2         1       1

In [39]:
X_test.head()

buying  maint  doors  persons  lug_boot  safety
467        4      2      2        2         3       1
617        4      1      1        1         2       1
229        2      4      4        2         1       3
1473       1      1      1        2         2       2
872        3      2      4        2         3       1

# Decision Tree Classifier with criterion gini index

In [41]:
from sklearn.tree import DecisionTreeClassifier
#the DecisionTreeClassifier model with criterion gini index

clf_gini = DecisionTreeClassifier(criterion='gini', max_depth=3, random_state=0)


# fit the model
clf_gini.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=3, random_state=0)

# Predict the Test set results with criterion gini index

In [45]:
y_test_pred_gini = clf_gini.predict(X_test)

# Check accuracy score with criterion gini index


Here, y_test are the true class labels and y_pred_gini are the predicted class labels in the test-set.

In [46]:
from sklearn.metrics import accuracy_score

print('Model accuracy score with criterion gini index: {0:0.4f}'. format(accuracy_score(y_test, y_test_pred_gini)))

Model accuracy score with criterion gini index: 0.7772


# Compare the train-set and test-set accuracy

Compare the train-set and test-set accuracy

Now, I will compare the train-set and test-set accuracy to check for overfitting.


In [48]:
y_train_pred_gini = clf_gini.predict(X_train)
y_train_pred_gini

array(['unacc', 'unacc', 'acc', ..., 'unacc', 'acc', 'unacc'],
      dtype=object)

In [50]:
print('Model accuracy score with criterion gini index: {0:0.4f}'. format(accuracy_score(y_train, y_train_pred_gini)))

Model accuracy score with criterion gini index: 0.7926


# Check for overfitting and underfitting

Here, The train_set accuracy is 0.7926 while The test_set accuracy is 0.7772. These two values to comparable.there is no overfitting.

In [53]:
# print the scores on training and test set

print('Training set score: {:.4f}'.format(clf_gini.score(X_train, y_train)))

print('Test set score: {:.4f}'.format(clf_gini.score(X_test, y_test)))

Training set score: 0.7926
Test set score: 0.7772


# Decision Tree Classifier with criterion entropy

In [54]:
clf_en = DecisionTreeClassifier(criterion='entropy', max_depth=3, random_state=0)


# fit the model
clf_en.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=3, random_state=0)

# Predict the Test set results with criterion entropy

In [56]:
y_test_pred_en = clf_en.predict(X_test)

# Check accuracy score with criterion entropy


In [57]:
from sklearn.metrics import accuracy_score

print('Model accuracy score with criterion entropy: {0:0.4f}'. format(accuracy_score(y_test, y_test_pred_en)))

Model accuracy score with criterion entropy: 0.7772


# Compare the train-set and test-set accuracy

Now, I will compare the training_set and test_set accuracy to check for overfitting

In [58]:
y_train_pred_en = clf_en.predict(X_train)
y_train_pred_en

array(['unacc', 'unacc', 'acc', ..., 'unacc', 'acc', 'unacc'],
      dtype=object)

In [59]:
print('Model accuracy score with criterion entropy: {0:0.4f}'. format(accuracy_score(y_train, y_train_pred_en)))

Model accuracy score with criterion entropy: 0.7926


# Check for overfitting and underfitting

In [60]:
# print the scores on training and test set

print('Training set score: {:.4f}'.format(clf_en.score(X_train, y_train)))

print('Test set score: {:.4f}'.format(clf_en.score(X_test, y_test)))

Training set score: 0.7926
Test set score: 0.7772


# 
We can see that the training-set score and test-set score is same as above. The training-set accuracy score is0.7926 while the test-set accuracy to be  0.7772. These two values are comparable. So, there is no sign of overfitting.

Now, based on the above analysis we can conclude that our classification model accuracy is very good. Our model is doing a very good job in terms of predicting the class labels.



# Confusion matrix

In [62]:
# Print the Confusion Matrix and slice it into four pieces

from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_test_pred_en)

print('Confusion matrix\n\n', cm)

Confusion matrix

 [[ 70   0  55   0]
 [ 24   0   0   0]
 [ 21   0 373   0]
 [ 24   0   3   0]]


# Classification Report

In [64]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_test_pred_en))

              precision    recall  f1-score   support

         acc       0.50      0.56      0.53       125
        good       0.00      0.00      0.00        24
       unacc       0.87      0.95      0.90       394
       vgood       0.00      0.00      0.00        27

    accuracy                           0.78       570
   macro avg       0.34      0.38      0.36       570
weighted avg       0.71      0.78      0.74       570



C:\Users\ASHOK\anaconda\lib\site-packages\sklearn\metrics\_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ASHOK\anaconda\lib\site-packages\sklearn\metrics\_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ASHOK\anaconda\lib\site-packages\sklearn\metrics\_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#  Results and conclusion

# 
In this project, I build a Decision-Tree Classifier model to predict the safety of the car. I build two models, one with criterion gini index and another one with criterion entropy. The model yields a very good performance as indicated by the model accuracy in both the cases which was found to be 0.7926.


In the model with criterion gini index, the training-set accuracy score is 0.7926 while the test-set accuracy to be 0.7772. These two values are quite comparable. So, there is no sign of overfitting.


Similarly, in the model with criterion entropy, the training-set accuracy score is 0.7926 while the test-set accuracy to be 0.7772.We get the same values as in the case with criterion gini. So, there is no sign of overfitting.


In both the cases, the training-set and test-set accuracy score is the same. It may happen because of small dataset.
The confusion matrix and classification report yields very good model performance.